include/lsst/afw/geom/SpanSet.h

```diff
@@ -0,0 +1,252 @@
+    typedef value_type const & const_reference;
+    const_iterator begin() const;
+    const_iterator end() const;
+    const_iterator cbegin() const;
+    const_iterator cend() const;
+    const_reference front() const;
+    const_reference back() const;
+    size_type size() const;
+    bool empty() const;
+
+    /// Default constructor creates a null set, useful as a placeholder and for the move constructor
+    SpanSet();
+
+    /// Construct a SpanSet from a region defined by a box
+    explicit SpanSet(Box2I const & box);
+
+    /// Construct a SpanSet from a container of Spans
+    template <typename iter>
+    SpanSet(iter begin, iter end, bool normalize = true):spanVector(begin, end) {
+        if (normalize) {
+            runNormalize();
+        }
+        initialize();
+    }
+
+    /// Construct a SpanSet from a std vector by copying
+    explicit SpanSet(const std::vector<Span> & vec, bool normalize = true);
+
+    /// Construct a SpanSet by moving the contents of a vector of Spans
+    explicit SpanSet(std::vector<Span> && vec, bool normalize = true);
+
+    /// Define class methods
+    size_type getArea() const;
+
+    Box2I getBBox() const;
+
+    bool isContiguous() const;
+
+    SpanSet shiftedBy(int x, int y) const;
+    SpanSet shiftedBy(Extent2I const & offset) const;
+
+    SpanSet clippedTo(Box2I const & box) const;
+
+    SpanSet transformedBy(LinearTransform const & t) const;
+    SpanSet transformedBy(AffineTransform const & t) const;
+    SpanSet transformedBy(XYTransform const & t) const;
+
+    bool overlaps(SpanSet const & other) const;
+
+    bool contains(SpanSet const & other) const;
+    bool contains(Point2I const & point) const;
+
+    Point2D computeCentroid() const;
+
+    ellipses::Quadrupole computeShape() const;
+
+    SpanSet dilate(int r, Stencil s = Stencil::CIRCLE) const;
+    SpanSet dilate(SpanSet const & other) const;
+
+    SpanSet erode(int r, Stencil s = Stencil::CIRCLE) const;
+    SpanSet erode(SpanSet const & other) const;
+
+    template <typename Pixel, int inC>
+    ndarray::Array<Pixel, 1, 1> flatten(ndarray::Array<Pixel, 2, inC>  & input,
+                                                       Point2I const xy0 = Point2I()) const {
+        // Populate a one dimensional array with the values from input at taken at the points of SpanSet
+        ndarray::Array<Pixel, 1, 1> outputArray = ndarray::allocate(ndarray::makeVector(getArea()));
+        outputArray.deep() = 0;
+        flatten(outputArray, input, xy0);
+        return outputArray;
+    }
+
+    template <typename Pixel, int outC, int inC>
+    void flatten(
+            ndarray::Array<Pixel, 1, outC> & output,
+            ndarray::Array<Pixel, 2, inC> & input,
+            Point2I const xy0 =  Point2I()) const {
+        // Populate array output with values from input at positions given by SpanSet
+        applyFunctor([](Point2I point, int n, typename details::FlatNdGetter<Pixel>::Reference out,
```

Comment:
True, but passing by reference is still good practice:
> Even when small objects have inexpensive copy constructors, there can be performance issues. Some compilers treat built-in and user-defined types differently, even if they have the same underlying representation. For example, some compilers refuse to put objects consisting of only a `double` into a register, even though they happily place naked `doubles` there on a regular basis. When that kind of thing happens, you can be better off passing such objects by reference, because compilers will certainly put pointers (the implementation of references) into registers.
-- Effective C++, 3rd Ed., p. 89